hasher-matcher-actioner/scripts/submit_s3_bucket

```diff
@@ -0,0 +1,142 @@
+#! /usr/bin/env python3
+# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
+
+"""
+Given an S3 bucket, and enough permissions on the credentials, iterates through
+all items and submits them as presigned URLs to HMA.
+
+Run without arguments for help.
+"""
+
+import uuid
+import argparse
+import typing as t
+import boto3
+import functools
+import concurrent.futures
+
+from requests.adapters import HTTPAdapter
+from requests.packages.urllib3.util.retry import Retry
+
+from script_utils import HasherMatcherActionerAPI
+
+
+@functools.lru_cache(maxsize=None)
+def _get_s3_client():
+    return boto3.client("s3")
+
+
+@functools.lru_cache(maxsize=None)
+def _get_hma_api(api_url, token):
+    retry_strategy = Retry(
+        total=5,
+        backoff_factor=1,
+        status_forcelist=[429, 500, 502, 503, 504],
+        allowed_methods=[
+            "GET",
+            "PUT",
+            "POST",
+        ],  # Including POST as the API should not perform an insert if an error is returned
+    )
+    api = HasherMatcherActionerAPI(
+        api_url, token, transport_adapter=HTTPAdapter(max_retries=retry_strategy)
+    )
+    return api
+
+
+class S3BucketSubmitter:
+    def __init__(
+        self,
+        bucket: str,
+        api: HasherMatcherActionerAPI,
+        prefix: t.Optional[str] = None,
+        loop: bool = False,
+    ):
+        self.bucket = bucket
+        self.prefix = prefix
+        self.api = api
+        self.loop = loop
+
+    def map_object_presigned_urls(self, callable: t.Callable):
+        paginator = _get_s3_client().get_paginator("list_objects_v2")
+        pages = paginator.paginate(Bucket=self.bucket, Prefix=self.prefix)
+
+        for page in pages:
+            for object in page["Contents"]:
+                if object["Size"] == 0:
+                    # Ignore folders and empty files
+                    continue
+
+                presigned_url = _get_s3_client().generate_presigned_url(
+                    "get_object",
+                    Params={
+                        "Bucket": self.bucket,
+                        "Key": object["Key"],
+                    },
+                    ExpiresIn=3 * 60 * 60,  # 3 hours from now.
+                )
+                callable(presigned_url)
+
+    def run(self):
+        if self.loop:
```

Comment:
Had to read through multiple times can you add a comment?

if loop=true does this just mean use multiple workers? Do they somehow share a  `paginator` how are repeats avoided?

Reply:
Added a comment. Repeats are not avoided. Loop merely causes the process to be repeated until Ctrl-Cd. We can't do 4000 QPS with 5000 images, so we loop and submit the same images again and again.

Reply:
got it, I'd probably recommends renaming loop to storm then but fine either way.